packages/protocol/contracts-0.8/common/EpochManager.sol

```diff
@@ -0,0 +1,338 @@
+    return currentEpochNumber;
+  }
+
+  /// returns epoch processing state
+  function getEpochProcessingState()
+    external
+    view
+    returns (uint256, uint256, uint256, uint256, uint256)
+  {
+    return (
+      uint256(epochProcessing.status),
+      epochProcessing.perValidatorReward,
+      epochProcessing.totalRewardsVoter,
+      epochProcessing.totalRewardsCommunity,
+      epochProcessing.totalRewardsCarbonFund
+    );
+  }
+
+  function getElected() external view returns (address[] memory) {
+    return elected;
+  }
+
+  function getFirstBlockAtEpoch(uint256 epoch) external view returns (uint256) {
+    require(epoch >= firstKnownEpoch, "Epoch not known");
+    require(epoch <= currentEpochNumber, "Epoch not created yet");
+    return epochs[epoch].firstBlock;
+  }
+
+  function getLastBlockAtEpoch(uint256 epoch) external view returns (uint256) {
+    require(epoch >= firstKnownEpoch, "Epoch not known");
+    require(epoch < currentEpochNumber, "Epoch not finished yet");
+    return epochs[epoch].lastBlock;
+  }
+
+  function isBlocked() external view returns (bool) {
+    return isOnEpochProcess();
+  }
+
+  /**
+   * @notice Returns the storage, major, minor, and patch version of the contract.
+   * @return Storage version of the contract.
+   * @return Major version of the contract.
+   * @return Minor version of the contract.
+   * @return Patch version of the contract.
+   */
+  function getVersionNumber() external pure returns (uint256, uint256, uint256, uint256) {
+    return (1, 1, 0, 0);
+  }
+
+  /**
+   * @notice Sets the time duration of an epoch.
+   * @param newEpochDuration The duration of an epoch in seconds.
+   * @dev Can only be set by owner.
+   */
+  function setEpochDuration(uint256 newEpochDuration) public onlyOwner {
+    epochDuration = newEpochDuration;
+  }
+
+  function isTimeForNextEpoch() public view returns (bool) {
+    return block.timestamp >= epochs[currentEpochNumber].startTimestamp + epochDuration;
+  }
+
+  function isOnEpochProcess() public view returns (bool) {
+    return epochProcessing.status == EpochProcessStatus.Started;
+  }
+
+  function systemAlreadyInitialized() public view returns (bool) {
+    return initialized && elected.length > 0;
+  }
+
+  function allocateValidatorsRewards() internal {
+    uint256 totalRewards = 0;
+    IScoreReader scoreReader = getScoreReader();
+    IValidators validators = getValidators();
+
+    for (uint i = 0; i < elected.length; i++) {
+      uint256 validatorScore = scoreReader.getValidatorScore(elected[i]);
+      uint256 validatorReward = validators.computeEpochReward(
+        elected[i],
+        validatorScore,
+        epochProcessing.perValidatorReward
+      );
+      validatorPendingPayments[elected[i]] += validatorReward;
+      totalRewards += validatorReward;
+    }
+    // Mint all cUSD required for payment and the corresponding CELO
+    validators.mintStableToEpochManager(totalRewards);
```

Comment:
why is the validator minting and not us?

To me, as EpochManager is the one minting CELO to compensate for the cUSD minted, then it should be doing both action, not only one

Reply:
the reason is that Validators.sol is already whitelisted on Mento tokens for minting, and it's hardcoded. So to avoid having to upgrade Mento contracts and coordinate with them, better to just proxy the call using validators.

Reply:
addressed by https://github.com/celo-org/celo-monorepo/pull/11223 commit https://github.com/celo-org/celo-monorepo/commit/66c4a89542ac3b03a3dad2fadcac803f7bba3205